### PR TITLE
JGroups failure detection configuration

### PR DESCRIPTION
**Pull request by pruivo**
Needs to be moved to a proper place...

### CODE REVIEW

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+:description: This guide describes how JGroups detects failed instances and its configuration.
+
+{description}
+
+== Audience
+
+This guide describe the failure detection mechanism that JGroups enforces to detect crashed pods inside the Infinispan cluster.
+It explains the configuration attributes available for customization and its trade-off to achieve the desired detection rate.
+
+== Broken TCP connections
```
Comment (pruivo) [inline doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:11]: @belaban when you get a chance, can you take a look at this section

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+However, it also increases the network usage due to the increase frequency of heartbeats.
+Note that all Infinispan pods in both cluster will send heartbeats to all the configured Gossip Router pods.
+
+TIP: The defaults should be work for 99% of the uses cases.
+
+For intra cluster communication in Infinispan or Keycloak cluster, broken TCP connection are detected and closed by the failure detection.
+See section <<fd>> for more information.
+
+[[fd]]
+== Failure Detection
```
Comment (pruivo) [inline doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:57]: @belaban when you get a chance, can you take a look at this section

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 1 above)
Comment (belaban) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:11]: * "have a performance impact" -> "has a ..."
* "estabilished"
* "The timeouts are tuning..."?

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 2 above)
Comment (belaban) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:57]: * FD_ALL3: perhaps mention that doesn't send out heartbeats when regular messages have been sent

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 1 above)
Comment (belaban) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:11]: LGTM!

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 2 above)
Comment (pruivo) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:57]: We use mostly distributed caches (use unicast only) and heartbeats are suppressed only with multicast:

https://github.com/belaban/JGroups/blob/901197b381b6228eb3d2fde862972839812dfce6/src/org/jgroups/protocols/FailureDetection.java#L338

Is it worth mentioning?

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 1 above)
Comment (pruivo) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:11]: Thanks for the comments!

File: doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc (hunk 2 above)
Comment (belaban) [reply doc/kubernetes/modules/ROOT/pages/running/timeout_tunning.adoc:57]: 'Multicast' in this respect means that we're sending a heartbeat to *everyone*, so in TCP this is an N-1 operation

**Comment by ahus1** [comment]
@pruivo - I did some editing for style and grammar, and also added a paragraph here and there describing the "why" of failure detection, and the impacts the different failures have. 

I read the tip you're providing in the docs:

> The defaults should work for 99% of the use cases.

I agree that this is a level a regular admin should not touch in those setups we want to support. The installation we describe should come with good defaults for a low latency within the cluster, as well as <10 ms latencies between the data centers. With all operations being in-memory, the only erratic behavior would be a garbage collection during normal operations.

I am hesitant to put this level of detail to the main Keycloak documentation, to avoid people changing those defaults when they shouldn't.  Therefore I want to keep them within the KCB documentation site for now. We might then reference and refine it when we have questions from the community around this. 

So once you think the documentation is ready from your end, switch this PR to ready for review and I'll merge it after a final review. 

Something we might need for customer and the community is something similar, but with a different perspective: What should be metrics or log messages that around timeouts that I should watch out for, and what would be possible causes and remedies. Still, this might be a different page and too much for this page, and it might grow as people ask questions around that. 

Please let me know you think differently.

**Comment by pruivo** [comment]
I've reviewed and squashed the commits. I'm fine keeping this outside the main Keycloak documentation. It should be available somewhere since it may help solution architects or support. 

Timeouts are usually a signal of trouble:

* Long GC pauses (miss configuration of the heap? memory leak?)
* Thread pool exhaustion (higher workload than expected?)
* Faulty network leading to dropped packages
* Pods crashing, requests will timeout until JGroups sends a new view change event
* Issues in the Keycloak side, because the `ClientListeners` are invoked in sync mode and with the lock acquired.

If we start thinking about it, we have material to write a book :smile:

**Pull request merged**